# **Waze Project**
**Course 1 - Foundations of Data Science**

Welcome to the Waze Project!

Your Waze data analytics team is still in the early stages of their user churn project. You have received notice that your team has been given access to Waze's user data. To get clear insights, the user data must be inspected and prepared for the upcoming process of exploratory data analysis (EDA).

A Python notebook has been prepared to guide you through this project. Answer the questions and create an executive summary for the Waze data team.

# **Course 1 End-of-course project: Inspect and analyze data**

In this activity, you will examine data provided and prepare it for analysis. This activity will help ensure the information is,

1.   Ready to answer questions and yield insights

2.   Ready for visualizations

3.   Ready for future hypothesis testing and statistical methods
<br/>

**The purpose** of this project is to investigate and understand the data provided.

**The goal** is to use a dataframe contructed within Python, perform a cursory inspection of the provided dataset, and inform team members of your findings.
<br/>

*This activity has three parts:*

**Part 1:** Understand the situation
* How can you best prepare to understand and organize the provided information?

**Part 2:** Understand the data

* Create a pandas dataframe for data learning, future exploratory data analysis (EDA), and statistical activities

* Compile summary information about the data to inform next steps

**Part 3:** Understand the variables

* Use insights from your examination of the summary data to guide deeper investigation into variables


<br/>

Follow the instructions and answer the following questions to complete the activity. Then, you will complete an Executive Summary using the questions listed on the PACE Strategy Document.

Be sure to complete this activity before moving on. The next course item will provide you with a completed exemplar to compare to your own work.



# **Identify data types and compile summary information**


<img src="images/Pace.png" width="100" height="100" align=left>

# **PACE stages**

Throughout these project notebooks, you'll see references to the problem-solving framework, PACE. The following notebook components are labeled with the respective PACE stages: Plan, Analyze, Construct, and Execute.

<img src="images/Plan.png" width="100" height="100" align=left>


## **PACE: Plan**

Consider the questions in your PACE Strategy Document and those below to craft your response:

### **Task 1. Understand the situation**

*   How can you best prepare to understand and organize the provided driver data?


*Begin by exploring your dataset and consider reviewing the Data Dictionary.*

i should use pandas to explore the columns and their data types. i will try exploring the max,min and medians of different coumns and see what the data shows me. 

<img src="images/Analyze.png" width="100" height="100" align=left>

## **PACE: Analyze**

Consider the questions in your PACE Strategy Document to reflect on the Analyze stage.

### **Task 2a. Imports and data loading**

Start by importing the packages that you will need to load and explore the dataset. Make sure to use the following import statements:

*   `import pandas as pd`

*   `import numpy as np`


In [2]:
# Import packages for data manipulation
import pandas as pd
import numpy as np

Then, load the dataset into a dataframe. Creating a dataframe will help you conduct data manipulation, exploratory data analysis (EDA), and statistical activities.

**Note:** As shown in this cell, the dataset has been automatically loaded in for you. You do not need to download the .csv file, or provide more code, in order to access the dataset and proceed with this lab. Please continue with this activity by completing the following instructions.

In [3]:
# Load dataset into dataframe
df = pd.read_csv('waze_dataset.csv')

### **Task 2b. Summary information**

View and inspect summary information about the dataframe by **coding the following:**

1.   df.head(10)
2.   df.info()

*Consider the following questions:*

1. When reviewing the `df.head()` output, are there any variables that have missing values?

2. When reviewing the `df.info()` output, what are the data types? How many rows and columns do you have?

3. Does the dataset have any missing values?

In [4]:
##inspect first 10
df.head(10)

,ID,label,sessions,drives,total_sessions,n_days_after_onboarding,total_navigations_fav1,total_navigations_fav2,driven_km_drives,duration_minutes_drives,activity_days,driving_days,device
0,0,retained,283,226,296.748273,2276,208,0,2628.845068,1985.775061,28,19,Android
1,1,retained,133,107,326.896596,1225,19,64,13715.920550,3160.472914,13,11,iPhone
2,2,retained,114,95,135.522926,2651,0,0,3059.148818,1610.735904,14,8,Android
3,3,retained,49,40,67.589221,15,322,7,913.591123,587.196542,7,3,iPhone
4,4,retained,84,68,168.247020,1562,166,5,3950.202008,1219.555924,27,18,Android
5,5,retained,113,103,279.544437,2637,0,0,901.238699,439.101397,15,11,iPhone
6,6,retained,3,2,236.725314,360,185,18,5249.172828,726.577205,28,23,iPhone
7,7,retained,39,35,176.072845,2999,0,0,7892.052468,2466.981741,22,20,iPhone
8,8,retained,57,46,183.532018,424,0,26,2651.709764,1594.342984,25,20,Android
9,9,churned,84,68,244.802115,2997,72,0,6043.460295,2341.838528,7,3,iPhone


In [5]:
###info of the data
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14999 entries, 0 to 14998
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   ID                       14999 non-null  int64  
 1   label                    14299 non-null  object 
 2   sessions                 14999 non-null  int64  
 3   drives                   14999 non-null  int64  
 4   total_sessions           14999 non-null  float64
 5   n_days_after_onboarding  14999 non-null  int64  
 6   total_navigations_fav1   14999 non-null  int64  
 7   total_navigations_fav2   14999 non-null  int64  
 8   driven_km_drives         14999 non-null  float64
 9   duration_minutes_drives  14999 non-null  float64
 10  activity_days            14999 non-null  int64  
 11  driving_days             14999 non-null  int64  
 12  device                   14999 non-null  object 
dtypes: float64(3), int64(8), object(2)
memory usage: 1.5+ MB


==> ENTER YOUR RESPONSES TO QUESTIONS 1-3 HERE
1. Noticeable missing values are total navs to a favorite destination. It has an assumptino of a favorite place or reoccuring routes. 
2. There are 13 columns and 14,299 rows. types of data are objects and numbers and a few floats. 
3. seems all the ccolumns have values in them even if it is a 0 value. 


### **Task 2c. Null values and summary statistics**

Compare the summary statistics of the 700 rows that are missing labels with summary statistics of the rows that are not missing any values.

**Question:** Is there a discernible difference between the two populations?


In [6]:
# Isolate rows with null values
##df[df.is null().any(columns)]
any_nulls = df[df.isnull().any(axis=1)]
# Display summary stats of rows with null values
print(any_nulls)


          ID label  sessions  drives  total_sessions  n_days_after_onboarding  \
77        77   NaN        63      50      133.104155                      783   
80        80   NaN       116      93      436.060183                     1584   
98        98   NaN        78      64      583.492789                     3414   
111      111   NaN       106     102      113.379056                     2228   
142      142   NaN        32      26      222.129310                      208   
...      ...   ...       ...     ...             ...                      ...   
14941  14941   NaN       191     160      485.328204                     1287   
14943  14943   NaN        48      38       96.797017                      555   
14945  14945   NaN        34      29      134.416604                     1643   
14972  14972   NaN       220     181      256.212166                     1718   
14993  14993   NaN        67      57       97.570074                     1131   

       total_navigations_fa

In [18]:
# Isolate rows without null values
##nonulls = df[df.notnull().any(axis=1)]
no_nulls = df[df.notnull().any(axis=1)]

# Display summary stats of rows without null values
print(no_nulls)

          ID     label  sessions  drives  total_sessions  \
0          0  retained       283     226      296.748273   
1          1  retained       133     107      326.896596   
2          2  retained       114      95      135.522926   
3          3  retained        49      40       67.589221   
4          4  retained        84      68      168.247020   
...      ...       ...       ...     ...             ...   
14994  14994  retained        60      55      207.875622   
14995  14995  retained        42      35      187.670313   
14996  14996  retained       273     219      422.017241   
14997  14997   churned       149     120      180.524184   
14998  14998  retained        73      58      353.419797   

       n_days_after_onboarding  total_navigations_fav1  \
0                         2276                     208   
1                         1225                      19   
2                         2651                       0   
3                           15                 

Seems the unlabeled ones are similar, but the labe of being churned is going to show more of a drop off and retained is a labeled. The total distance seems higher for those retained though which shows positive trends with regular users. Also activity numbers vary in a more positive way when they are labeled. 

### **Task 2d. Null values - device counts**

Next, check the two populations with respect to the `device` variable.

**Question:** How many iPhone users had null values and how many Android users had null values?

In [20]:
# Get count of null values by device
##use premade any null list[devicecolumn].countit
any_nulls['device'].value_counts()

iPhone     447
Android    253
Name: device, dtype: int64

iphone had 447 and android has 253

Now, of the rows with null values, calculate the percentage with each device&mdash;Android and iPhone. You can do this directly with the [`value_counts()`](https://pandas.pydata.org/docs/reference/api/pandas.Series.value_counts.html) function.

In [22]:
# Calculate % of iPhone nulls and Android nulls
##nullslist[device column].valuecount(normalize)
any_nulls['device'].value_counts(normalize=True)


iPhone     0.638571
Android    0.361429
Name: device, dtype: float64

How does this compare to the device ratio in the full dataset?

In [23]:
# Calculate % of iPhone users and Android users in full dataset
##full data set[column device]. value_count(normalize=true)
df['device'].value_counts(normalize=True)

iPhone     0.644843
Android    0.355157
Name: device, dtype: float64

The percentage of missing values by each device is consistent with their representation in the data overall.

There is nothing to suggest a non-random cause of the missing data.

Examine the counts and percentages of users who churned vs. those who were retained. How many of each group are represented in the data?

In [25]:
# Calculate counts of churned vs. retained
##data.groubpby(labeltype)[devicecolumn].value_counts(normalize true to make percent)
df.groupby('label')['device'].value_counts(normalize=True)


label     device 
churned   iPhone     0.648659
          Android    0.351341
retained  iPhone     0.644393
          Android    0.355607
Name: device, dtype: float64

This dataset contains 82% retained users and 18% churned users.

Next, compare the medians of each variable for churned and retained users. The reason for calculating the median and not the mean is that you don't want outliers to unduly affect the portrayal of a typical user. Notice, for example, that the maximum value in the `driven_km_drives` column is 21,183 km. That's more than half the circumference of the earth!

In [27]:
# Calculate median values of all columns for churned and retained users
##data.groupby('labelcolumn').median(numeric_only=True)
df.groupby('label').median(numeric_only=True)

,ID,sessions,drives,total_sessions,n_days_after_onboarding,total_navigations_fav1,total_navigations_fav2,driven_km_drives,duration_minutes_drives,activity_days,driving_days
label,,,,,,,,,,,
churned,7477.5,59.0,50.0,164.339042,1321.0,84.5,11.0,3652.655666,1607.183785,8.0,6.0
retained,7509.0,56.0,47.0,157.586756,1843.0,68.0,9.0,3464.684614,1458.046141,17.0,14.0


This offers an interesting snapshot of the two groups, churned vs. retained:

Users who churned averaged ~3 more drives in the last month than retained users, but retained users used the app on over twice as many days as churned users in the same time period.

The median churned user drove ~200 more kilometers and 2.5 more hours during the last month than the median retained user.

It seems that churned users had more drives in fewer days, and their trips were farther and longer in duration. Perhaps this is suggestive of a user profile. Continue exploring!

Calculate the median kilometers per drive in the last month for both retained and churned users.

Begin by dividing the `driven_km_drives` column by the `drives` column. Then, group the results by churned/retained and calculate the median km/drive of each group.

In [28]:
# Add a column to df called `km_per_drive`
#data[km new column] = data[driven km] / data[drives]
df['km_per_drive'] = df['driven_km_drives'] / df['drives']

# Group by `label`, calculate the median, and isolate for km per drive
###data.groupby('label')[[new kn colum]].median()
df.groupby('label')[['km_per_drive']].median()

,km_per_drive
label,
churned,74.109416
retained,75.014702


The median retained user drove about one more kilometer per drive than the median churned user. How many kilometers per driving day was this?

To calculate this statistic, repeat the steps above using `driving_days` instead of `drives`.

In [30]:
# Add a column to df called `km_per_driving_day`
#data[km-perday] = data[driven_km_drives] / data['driving_days']
df['km_per_driving_day'] = df['driven_km_drives'] / df['driving_days'] 
# Group by `label`, calculate the median, and isolate for km per driving day
##data.groupby(label')[[km_per_driving_day]].median()
df.groupby('label')[['km_per_driving_day']].median()

,km_per_driving_day
label,
churned,697.541999
retained,289.549333


Now, calculate the median number of drives per driving day for each group.

In [32]:
# Add a column to df called `drives_per_driving_day`
##data[driveperdrivingday new column] = data[drives] / data [driving days]
df['drives_per_driving_day'] = df['drives'] / df['driving_days']

# Group by `label`, calculate the median, and isolate for drives per driving day
##data.groupby(label')[[drivesperdrivingday new column]].median()
df.groupby('label')[['drives_per_driving_day']].median()

,drives_per_driving_day
label,
churned,10.0000
retained,4.0625


The median user who churned drove 698 kilometers each day they drove last month, which is almost ~240% the per-drive-day distance of retained users. The median churned user had a similarly disproporionate number of drives per drive day compared to retained users.

It is clear from these figures that, regardless of whether a user churned or not, the users represented in this data are serious drivers! It would probably be safe to assume that this data does not represent typical drivers at large. Perhaps the data&mdash;and in particular the sample of churned users&mdash;contains a high proportion of long-haul truckers.

In consideration of how much these users drive, it would be worthwhile to recommend to Waze that they gather more data on these super-drivers. It's possible that the reason for their driving so much is also the reason why the Waze app does not meet their specific set of needs, which may differ from the needs of a more typical driver, such as a commuter.

Finally, examine whether there is an imbalance in how many users churned by device type.

Begin by getting the overall counts of each device type for each group, churned and retained.

In [33]:
# For each label, calculate the number of Android users and iPhone users
# Calculate counts of churned vs. retained
##data.groubpby(labeltype)[devicecolumn].value_counts()
df.groupby('label')['device'].value_counts()


label     device 
churned   iPhone     1645
          Android     891
retained  iPhone     7580
          Android    4183
Name: device, dtype: int64

Now, within each group, churned and retained, calculate what percent was Android and what percent was iPhone.

In [34]:
# For each label, calculate the percentage of Android users and iPhone user
##data.groupby(label')['device'].valuecounts(normalize=True)
df.groupby('label')['device'].value_counts(normalize=True)

label     device 
churned   iPhone     0.648659
          Android    0.351341
retained  iPhone     0.644393
          Android    0.355607
Name: device, dtype: float64

The ratio of iPhone users and Android users is consistent between the churned group and the retained group, and those ratios are both consistent with the ratio found in the overall dataset.

<img src="images/Construct.png" width="100" height="100" align=left>

## **PACE: Construct**

**Note**: The Construct stage does not apply to this workflow. The PACE framework can be adapted to fit the specific requirements of any project.



<img src="images/Execute.png" width="100" height="100" align=left>

## **PACE: Execute**

Consider the questions in your PACE Strategy Document and those below to craft your response:

### **Task 3. Conclusion**

Recall that your supervisor, May Santer, asked you to share your findings with the data team in an executive summary. Consider the following questions as you prepare to write your summary. Think about key points you may want to share with the team, and what information is most relevant to the user churn project.

**Questions:**

1. Did the data contain any missing values? How many, and which variables were affected? Was there a pattern to the missing data? yes there were 700 un labled ones which didnt greatly effect the over all numbers but it was a good chunk of users that wouldnt be counted in the churned comparisons. 

2. What is a benefit of using the median value of a sample instead of the mean?
median shows the actual middle number where as the mean does an avergae function and can be altered more by high or low outliers. 

3. Did your investigation give rise to further questions that you would like to explore or ask the Waze team about?
Was there any report stats that could be compared with the churned numbers?

4. What percentage of the users in the dataset were Android users and what percentage were iPhone users?
it was roughly 65% apple 35% androif. 

5. What were some distinguishing characteristics of users who churned vs. users who were retained?
Definitely higher reoccuring numbers with retained. churned showed a decrease in trips
6. Was there an appreciable difference in churn rate between iPhone users vs. Android users?
They actually had similar rates, so the phone model i dont believe was a reason for churn rates. 




1. Did the data contain any missing values? How many, and which variables were affected? Was there a pattern to the missing data? yes there were 700 un labled ones which didnt greatly effect the over all numbers but it was a good chunk of users that wouldnt be counted in the churned comparisons. 

2. What is a benefit of using the median value of a sample instead of the mean?
median shows the actual middle number where as the mean does an avergae function and can be altered more by high or low outliers. 

3. Did your investigation give rise to further questions that you would like to explore or ask the Waze team about?
Was there any report stats that could be compared with the churned numbers?

4. What percentage of the users in the dataset were Android users and what percentage were iPhone users?
it was roughly 65% apple 35% androif. 

5. What were some distinguishing characteristics of users who churned vs. users who were retained?
Definitely higher reoccuring numbers with retained. churned showed a decrease in trips
6. Was there an appreciable difference in churn rate between iPhone users vs. Android users?
They actually had similar rates, so the phone model i dont believe was a reason for churn rates. 

**Congratulations!** You've completed this lab. However, you may not notice a green check mark next to this item on Coursera's platform. Please continue your progress regardless of the check mark. Just click on the "save" icon at the top of this notebook to ensure your work has been logged.